File: problem-solving/adventofcode2020/day2.py
# https://adventofcode.com/2020/day/2
input_file = open('./day2.txt', 'r')
input_list = input_file.read().splitlines()

def solve_first():
    valid_count = 0
    for input in input_list:
        times, letter, password = input.split()
        lowest_time, highest_time = list(map(int, times.split("-")))
        letter = letter[0]
        if check_if_password_has_valid_letter_count(password, letter, lowest_time, highest_time):
            valid_count += 1
    return valid_count

def check_if_password_has_valid_letter_count(password, letter, lowest_time, highest_time):
    count = password.count(letter)
    if count >= lowest_time and count <= highest_time:
        return True
    return False

def solve_second():
    valid_count = 0
    for input in input_list:
        position, letter, password = input.split()
        first_position, second_position = list(map(int, position.split("-")))
        letter = letter[0]
        if check_if_password_has_one_right_letter_position(password, letter, first_position, second_position):
            valid_count += 1
    return valid_count

def check_if_password_has_one_right_letter_position(password, letter, first_position, second_position):
    first_contains = password[first_position-1] == letter
    second_contains = password[second_position-1] == letter
    if first_contains and not second_contains or not first_contains and second_contains:
        return True
    return False

if __name__ == "__main__":
    print(solve_first())
    print(solve_second())

File: problem-solving/adventofcode2020/day2.cpp
// https://adventofcode.com/2020/day/2
#include <iostream>
#include <string>
#include <vector>
#include <fstream>
#include <cstring>

using namespace std;

struct Policy
{
    int low, high;
    char letter;
    string password;

    bool hasValidLetterCount()
    {
        int letterCount = count(password.c_str(), password.c_str() + strlen(password.c_str()), letter);
        if (low <= letterCount && letterCount <= high)
        {
            return true;
        }
        return false;
    }

    bool hasOneRightLetterPosition()
    {
        bool lowConains = (password[low - 1] == letter);
        bool highContains = (password[high - 1] == letter);
        return lowConains != highContains;
    }
};

vector<Policy> readInput(string filepath)
{
    vector<Policy> inputList;
    ifstream inputFile;
    inputFile.open(filepath);

    string ranges, letter, password;
    while (inputFile >> ranges >> letter >> password)
    {
        Policy policy;
        sscanf(ranges.c_str(), "%d-%d", &policy.low, &policy.high);
        sscanf(letter.c_str(), "%c:", &policy.letter);
        policy.password = password;
        inputList.push_back(policy);
    }
    inputFile.close();
    return inputList;
}

int solveFirst(vector<Policy> &inputList)
{
    int validCount = 0;
    for (int i = 0; i < inputList.size(); i++)
    {
        if (inputList[i].hasValidLetterCount())
        {
            validCount++;
        }
    }
    return validCount;
}

int solveSecond(vector<Policy> &inputList)
{
    int validCount = 0;
    for (int i = 0; i < inputList.size(); i++)
    {
        if (inputList[i].hasOneRightLetterPosition())
        {
            validCount++;
        }
    }
    return validCount;
}

int main()
{
    vector<Policy> inputList = readInput("./day2.txt");
    cout << solveFirst(inputList) << endl;
    cout << solveSecond(inputList) << endl;
}